src/mds/FSMap.cc

```diff
@@ -0,0 +1,762 @@
+{
+  DECODE_START(1, p);
+  ::decode(fscid, p);
+  bufferlist mdsmap_bl;
+  ::decode(mdsmap_bl, p);
+  bufferlist::iterator mdsmap_bl_iter = mdsmap_bl.begin();
+  mds_map.decode(mdsmap_bl_iter);
+  DECODE_FINISH(p);
+}
+
+int FSMap::parse_filesystem(
+      std::string const &ns_str,
+      std::shared_ptr<Filesystem> *result
+      ) const
+{
+  std::string ns_err;
+  fs_cluster_id_t fscid = strict_strtol(ns_str.c_str(), 10, &ns_err);
+  if (!ns_err.empty() || filesystems.count(fscid) == 0) {
+    for (auto fs : filesystems) {
+      if (fs.second->mds_map.fs_name == ns_str) {
+        *result = fs.second;
+        return 0;
+      }
+    }
+    return -ENOENT;
+  } else {
+    *result = get_filesystem(fscid);
+    return 0;
+  }
+}
+
+void Filesystem::print(std::ostream &out) const
+{
+  // TODO add a non-json print?
+  JSONFormatter f;
+  dump(&f);
+  f.flush(out);
+}
+
+mds_gid_t FSMap::find_standby_for(mds_role_t role, const std::string& name) const
+{
+  mds_gid_t result = MDS_GID_NONE;
+
+  // First see if we have a STANDBY_REPLAY
+  auto fs = get_filesystem(role.fscid);
+  for (const auto &i : fs->mds_map.mds_info) {
+    const auto &info = i.second;
+    if (info.rank == role.rank && info.state == MDSMap::STATE_STANDBY_REPLAY) {
+      return info.global_id;
+    }
+  }
+
+  // See if there are any STANDBY daemons available
+  for (const auto &i : standby_daemons) {
+    const auto &gid = i.first;
+    const auto &info = i.second;
+    assert(info.state == MDSMap::STATE_STANDBY);
+    assert(info.rank == MDS_RANK_NONE);
+
+    if (info.laggy()) {
+      continue;
+    }
+
+    if ((info.standby_for_rank == role.rank && info.standby_for_ns == role.fscid)
+        || (name.length() && info.standby_for_name == name)) {
+      // It's a named standby for *me*, use it.
+      return gid;
+    } else if (info.standby_for_rank < 0 && info.standby_for_name.length() == 0)
+      // It's not a named standby for anyone, use it if we don't find
+      // a named standby for me later.
+      result = gid;
+  }
+
+  return result;
+}
+
+mds_gid_t FSMap::find_unused(bool force_standby_active) const {
+  for (const auto &i : standby_daemons) {
+    const auto &gid = i.first;
+    const auto &info = i.second;
+    assert(info.state == MDSMap::STATE_STANDBY);
+
+    if (info.laggy() || info.rank >= 0)
+      continue;
+
+    if ((info.standby_for_rank == MDSMap::MDS_NO_STANDBY_PREF ||
+         info.standby_for_rank == MDSMap::MDS_MATCHED_ACTIVE ||
+         (info.standby_for_rank == MDSMap::MDS_STANDBY_ANY
+          && force_standby_active))) {
+      return gid;
+    }
+  }
+  return MDS_GID_NONE;
+}
+
+mds_gid_t FSMap::find_replacement_for(mds_role_t role, const std::string& name,
+                               bool force_standby_active) const {
+  const mds_gid_t standby = find_standby_for(role, name);
+  if (standby)
+    return standby;
+  else
+    return find_unused(force_standby_active);
+}
+
+void FSMap::sanity() const
+{
```

Comment:
Follow-up commit: I like this function but I think we'll want a config option to disable it in case somebody gets into big trouble with respawning MDS daemons or something.